plat: rpi5: add basic Raspberry Pi 5 support

lorc opened the pull request:
RPi5 is based on new BCM2712 SoC which is based on quad Cortex-A76.

BCM2712 still does not provide secure memory so we are free to locate OP-TEE anything we want. It would be most beneficial to locate OP-TEE right after TF-A, at address 0x80000, but RPi5 loader places kernel there and it's location can't be changed.

According to PCB silkscreen, RPi5 boards can have 1GB, 2GB, 4GB or 8GB of memory. To be compatible with any variant, OP-TEE is placed close to the end of the first gigabyte.

BCM2712 uses PL011 as debug UART so we enable its driver.

According to specification, BCM2712 includes cryptography extensions, but this basic port does not enable them.


---

Also I made corresponding changes to TF-A which I'll publish soon.

lorc commented:
Corresponding TF-A patch: https://review.trustedfirmware.org/c/TF-A/trusted-firmware-a/+/28990

jenswi-linaro commented on core/arch/arm/plat-rpi5/main.c:
```diff
@@ -0,0 +1,47 @@
+
+#include <console.h>
+#include <drivers/pl011.h>
+#include <kernel/panic.h>
```
Please trim the includes to only what's needed.

jforissier commented:
@lorc any chance you could address Jen's comment? This can easily get merged after that I think.

lorc commented:
Yeah, sorry. I am waiting complementary patch to ARM TF to be merged. Changes here may affect this PR.

I'll mark it as a draft for now.

jforissier commented on core/mm/core_mmu.c:
```diff
@@ -1651,7 +1651,9 @@ enum teecore_memtypes core_mmu_get_type_by_pa(paddr_t pa)
 	struct tee_mmap_region *map = find_map_by_pa(pa);
 
-	if (!map)
+	/* VA spaces are have no valid PAs in the memory map */
```
s/are have/have/

jforissier commented on core/arch/arm/plat-rpi5/main.c:
```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (c) 2016, Sequitur Labs Inc. All rights reserved.
+ * Copyright (c) 2024, EPAM Systems.
+ * All rights reserved.
```
All rights reserved is not desired. Please keep only the Copyright lines.

jforissier commented on core/arch/arm/plat-rpi5/main.c:
```diff
@@ -0,0 +1,47 @@
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
```
Remove the license text. It is implied by the SPDX-License-Identifier.

jforissier commented on core/arch/arm/plat-rpi5/platform_config.h:
```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright (c) 2016, Sequitur Labs Inc. All rights reserved.
+ * Copyright (c) 2024, EPAM Systems.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
```
Remove

jforissier reviewed (commented):
Hi @lorc,

With my comments addressed:
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

lorc commented:
Okay, so approach with TF-A sideloading failed because of RPI5 firmware changes. Best I can do right now is to use `OPTEE_ALLOW_SMC_LOAD=1` feature and load OP-TEE via Linux kernel or U-Boot. 

BTW, during working on OPTEE_LOAD I have found another issue which prevented me from using DT generated by TF-A. I added a second patch to this PR. But I am not sure that this is the correct approach.

Maybe it is better to move PAs for VA spaces to the very top? Like  place them at 0xFFFFFFFFFFFFFFFF, which is well out of range for aarch64 address space.

lorc commented:
@jforissier, thank you for the review. I addressed your comments.  Also I added rpi5 build to github CI workflow.

But as I said, I am not sure that I made the proper fix for DT issue.

jenswi-linaro commented:
Regarding "mmu: ignore VA spaces in core_mmu_get_type_by_pa", I think we have another fix for that upstream.

yogeshbdeshpande commented:
@lorc can you please update the missing dependencies this PR needs so that we can follow up (if required) so as to proceed this work ?

jforissier commented:
Since there is interest I am re-opening this PR.

lorc commented:
Hi @yogeshbdeshpande,

sorry for the late reply. The biggest problem is loading OP-TEE into memory. I have an ARM-TF patch that does this in a hacky way (see second comment in this PR). But this patch didn't got a traction. This is the only thing that prevents RPI5 support from OP-TEE side.

lorc commented:
Okay then, I'll need some time to get back the board. I'll update PR in a few days.

yogeshbdeshpande commented:
> Okay then, I'll need some time to get back the board. I'll update PR in a few days.

Thank you very much @lorc , look forward to hearing an update from you!

lorc commented:
Yep, sorry for the delay, but I am stuck in debugging very nasty issue  with nested GIC emulation in KVM... I'll get back to OP-TEE once I am finished with that.

lorc commented:
Hi @spark404, code-wise it is ready. All you need is to address that last Jerome's comment for code, rebase changes to the latest HEAD and test if it still works with that patch for ARM-TF. After that you can write documentation and integrate it into CI, as Jerome requested.
